src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -484,6 +489,7 @@ private TableEntry(OneOrMany<T> items, ImmutableArray<EntryState> states)
 
                 _items = items;
                 _states = states;
+                _anyRemoved = states.Contains(EntryState.Removed);
```

Comment:
Let's either just do the `states.Contains(...)` check when we call `AsCached()` or track the removal state when we're making the builder (set a flag as soon as `RemoveEntries()` is called and pass that in here)

Reply:
> track the removal state when we're making the builder (set a flag as soon as `RemoveEntries()` is called and pass that in here)

Not sure what exactly you mean, but tried computing the flag more efficiently.

> Let's either just do the `states.Contains(...)` check when we call `AsCached()`

That seems the simplest option, could do that instead.